lib/utils/CHIMETelescope.cpp

```diff
@@ -3,16 +3,121 @@
 #include "ICETelescope.hpp"
 #include "Telescope.hpp"
 
-
 REGISTER_TELESCOPE(CHIMETelescope, "chime");
 
-
 CHIMETelescope::CHIMETelescope(const kotekan::Config& config, const std::string& path) :
     ICETelescope(config.get<std::string>(path, "log_level")) {
 
-    // TODO: rename this parameter to `num_freq_per_stream` in the config
-    _num_freq_per_stream = config.get<uint32_t>(path, "num_local_freq");
+    // This is always 1 for CHIME
+    _num_freq_per_stream = 1;
+
+    _require_frequency_map = config.get_default<bool>(path, "require_frequency_map", false);
+    _allow_default_map = config.get_default<bool>(path, "allow_default_map", true);
 
     set_sampling_params(800.0, 2048, 2);
     set_gps(config, path);
+    set_frequency_map(config, "fpga_frequency_map");
+}
+
+void CHIMETelescope::set_frequency_map(const kotekan::Config& config, const std::string& path) {
+    nlohmann::json fpga_freq_map_json = config.get_default<nlohmann::json>("/", path, {});
+
+    if (fpga_freq_map_json.empty()) {
+        if (_require_frequency_map) {
+            std::string msg = fmt::format("Frequency map required by not found at /{:s}", path);
+            ERROR("{:s}", msg);
+            throw std::runtime_error(msg);
+        }
+
+        if (!_allow_default_map) {
+            WARN("The generation of a default frequency map is disabled, "
+                 "calls to convert stream IDs will generate an exception");
+            return;
+        }
+
+        WARN("No frequency map provided, generating default map");
+
+        // Default FPGA mapping will be incorrect for any dynamic map
+        // This function generates the default FPGA table mapping in the CHIME
+        // configuration 4 pairs of ICEBoards, won't work for any other config.
+        ice_stream_id_t stream_id = {0, 0, 0, 0};
+        for (uint32_t i = 0; i < nfreq; ++i) {
+
+            DEBUG2("Adding table entry: stream_id: {:d} : crate: {:d}, "
+                   "slot: {:d}, link: {:d}, unused: {:d} = {:d}",
+                   ice_encode_stream_id(stream_id).id, stream_id.crate_id, stream_id.slot_id,
+                   stream_id.link_id, stream_id.unused, i);
+            frequency_table[ice_encode_stream_id(stream_id).id] = i;
+
+            stream_id.slot_id = (stream_id.slot_id + 1) % 16;
+
+            // Only update slot_id on first pass.
+            if (i == 0)
+                continue;
+
+            if (i % 256 == 0) {
+                stream_id.unused = (stream_id.unused + 1) % 4;
+            }
+            if (i % 32 == 0) {
+                stream_id.link_id = (stream_id.link_id + 1) % 8;
+            }
+            if (i % 16 == 0) {
+                stream_id.crate_id = (stream_id.crate_id + 1) % 2;
+            }
```

Comment:
I find this logic hard to follow, with elements only incrementing conditionally, and only after the first round is set. Isn't it essentially equivalent to this:
```suggestion

            stream_id.slot_id = i % 16; // If I'm unsure about a part of this, it's this line
            stream_id.unused = (i / 256) % 4;
            stream_id.link_id = (i / 32) % 8;
            stream_id.crate_id = (i / 16) % 2;
            stream_id.unused = (i / 256) % 4;

            DEBUG2("Adding table entry: stream_id: {:d} : crate: {:d}, "
                   "slot: {:d}, link: {:d}, unused: {:d} = {:d}",
                   ice_encode_stream_id(stream_id).id, stream_id.crate_id, stream_id.slot_id,
                   stream_id.link_id, stream_id.unused, i);
            frequency_table[ice_encode_stream_id(stream_id).id] = i;
```

Reply:
I agree, this is much cleaner.  I've tested it, and it does produce the right values for all bin numbers.